File: tests/state-store.test.ts
import { mkdtemp, rm } from "node:fs/promises";
import os from "node:os";
import path from "node:path";

import { afterEach, describe, expect, it } from "vitest";

import { StateStore } from "../src/storage/state-store.js";

const tempDirs: string[] = [];

afterEach(async () => {
  await Promise.all(tempDirs.splice(0).map((dir) => rm(dir, { recursive: true, force: true })));
});

describe("StateStore", () => {
  it("persists config, session, and encrypted credentials", async () => {
    const baseDir = await createTempDir();
    const store = new StateStore(baseDir);

    await store.writeConfig({
      baseUrl: "https://docs.example.com",
      email: "jane@example.com",
      lastAuthenticatedAt: "2026-03-12T00:00:00.000Z",
    });
    await store.writeSession({
      token: "token-value",
      expiresAt: "2026-03-12T01:00:00.000Z",
      savedAt: "2026-03-12T00:00:00.000Z",
    });
    await store.writeCredentials({
      email: "jane@example.com",
      password: "super-secret",
    });

    await expect(store.readConfig()).resolves.toEqual({
      baseUrl: "https://docs.example.com",
      email: "jane@example.com",
      lastAuthenticatedAt: "2026-03-12T00:00:00.000Z",
    });
    await expect(store.readSession()).resolves.toEqual({
      token: "token-value",
      expiresAt: "2026-03-12T01:00:00.000Z",
      savedAt: "2026-03-12T00:00:00.000Z",
    });
    await expect(store.readCredentials()).resolves.toEqual({
      email: "jane@example.com",
      password: "super-secret",
    });
  });

  it("clears the saved session without touching credentials", async () => {
    const baseDir = await createTempDir();
    const store = new StateStore(baseDir);

    await store.writeSession({
      token: "token-value",
      expiresAt: null,
      savedAt: "2026-03-12T00:00:00.000Z",
    });
    await store.writeCredentials({
      email: "jane@example.com",
      password: "super-secret",
    });

    await store.clearSession();

    await expect(store.readSession()).resolves.toBeNull();
    await expect(store.readCredentials()).resolves.toEqual({
      email: "jane@example.com",
      password: "super-secret",
    });
  });
});

async function createTempDir(): Promise<string> {
  const dir = await mkdtemp(path.join(os.tmpdir(), "docmost-local-mcp-"));
  tempDirs.push(dir);
  return dir;
}

File: tests/state_store_test.rs
use docmost_local_mcp::{
    storage::state_store::StateStore,
    types::{StoredConfig, StoredCredentials, StoredSession},
};
use tempfile::TempDir;

#[tokio::test]
async fn persists_config_session_and_encrypted_credentials() {
    unsafe {
        std::env::set_var("DOCMOST_DISABLE_KEYRING", "1");
    }
    let temp_dir = TempDir::new().unwrap();
    let store = StateStore::new(Some(temp_dir.path().to_path_buf())).unwrap();

    store
        .write_config(&StoredConfig {
            base_url: "https://docs.example.com".to_string(),
            email: "jane@example.com".to_string(),
            last_authenticated_at: "2026-03-12T00:00:00.000Z".to_string(),
        })
        .await
        .unwrap();
    store
        .write_session(&StoredSession {
            token: "token-value".to_string(),
            expires_at: Some("2026-03-12T01:00:00.000Z".to_string()),
            saved_at: "2026-03-12T00:00:00.000Z".to_string(),
        })
        .await
        .unwrap();
    store
        .write_credentials(&StoredCredentials {
            email: "jane@example.com".to_string(),
            password: "super-secret".to_string(),
        })
        .await
        .unwrap();

    assert_eq!(
        store.read_config().await.unwrap(),
        Some(StoredConfig {
            base_url: "https://docs.example.com".to_string(),
            email: "jane@example.com".to_string(),
            last_authenticated_at: "2026-03-12T00:00:00.000Z".to_string(),
        })
    );
    assert_eq!(
        store.read_session().await.unwrap(),
        Some(StoredSession {
            token: "token-value".to_string(),
            expires_at: Some("2026-03-12T01:00:00.000Z".to_string()),
            saved_at: "2026-03-12T00:00:00.000Z".to_string(),
        })
    );
    assert_eq!(
        store.read_credentials().await.unwrap(),
        Some(StoredCredentials {
            email: "jane@example.com".to_string(),
            password: "super-secret".to_string(),
        })
    );
}

#[tokio::test]
async fn clears_saved_session_without_touching_credentials() {
    unsafe {
        std::env::set_var("DOCMOST_DISABLE_KEYRING", "1");
    }
    let temp_dir = TempDir::new().unwrap();
    let store = StateStore::new(Some(temp_dir.path().to_path_buf())).unwrap();

    store
        .write_session(&StoredSession {
            token: "token-value".to_string(),
            expires_at: None,
            saved_at: "2026-03-12T00:00:00.000Z".to_string(),
        })
        .await
        .unwrap();
    store
        .write_credentials(&StoredCredentials {
            email: "jane@example.com".to_string(),
            password: "super-secret".to_string(),
        })
        .await
        .unwrap();

    store.clear_session().await.unwrap();

    assert_eq!(store.read_session().await.unwrap(), None);
    assert_eq!(
        store.read_credentials().await.unwrap(),
        Some(StoredCredentials {
            email: "jane@example.com".to_string(),
            password: "super-secret".to_string(),
        })
    );
}
